File: lexer/lexer_test.go
package lexer

import (
	"jparse/token"
	"testing"
)

type lexerTestCase struct {
	input          string
	expectedTokens []token.Token
}

func TestNextToken(t *testing.T) {
	tests := []lexerTestCase{
		{
			input: `=+(),`,
			expectedTokens: []token.Token{
				{Kind: token.Bind, Literal: "="},
				{Kind: token.Plus, Literal: "+"},
				{Kind: token.LParen, Literal: "("},
				{Kind: token.RParen, Literal: ")"},
				{Kind: token.Comma, Literal: ","},
			},
		},
		{
			input: `//STEP010  EXEC PGM=SORT
//SYSOUT   DD SYSOUT=*
//SORTIN   DD DSN=userid.IBMMF.INPUT,DISP=SHR
//SORTOUT  DD DSN=userid.IBMMF.OUTPUT,DISP=SHR
//SYSIN    DD *
  SORT FIELDS=(1,15,ZD,A)
  SUM FIELDS=NONE
/*`,
			expectedTokens: []token.Token{
				{Kind: token.Executable, Literal: "//"},
				{Kind: token.Tag, Literal: "STEP010"},
				{Kind: token.Exec, Literal: "EXEC"},
				{Kind: token.Parameter, Literal: "PGM=SORT"},
				{Kind: token.Executable, Literal: "//"},
				{Kind: token.Tag, Literal: "SYSOUT"},
				{Kind: token.Dd, Literal: "DD"},
				{Kind: token.Parameter, Literal: "SYSOUT=*"},
				{Kind: token.Executable, Literal: "//"},
				{Kind: token.Tag, Literal: "SORTIN"},
				{Kind: token.Dd, Literal: "DD"},
				{Kind: token.Parameter, Literal: "DSN=userid.IBMMF.INPUT"},
				{Kind: token.Parameter, Literal: "DISP=shr"},
				{Kind: token.Executable, Literal: "//"},
				{Kind: token.Tag, Literal: "SORTOUT"},
				{Kind: token.Dd, Literal: "DD"},
				{Kind: token.Parameter, Literal: "DSN=userid.IBMMF.OUTPUT"},
				{Kind: token.Parameter, Literal: "DISP=shr"},
				{Kind: token.Executable, Literal: "//"},
				{Kind: token.Tag, Literal: "SYSIN"},
				{Kind: token.Dd, Literal: "DD"},
				{Kind: token.Aster, Literal: "*"},
				{Kind: token.SysinMarker, Literal: "/*"},
			},
		},
	}

	runTest(t, tests)
}

// runner for the test asserts
func runTest(t *testing.T, tests []lexerTestCase) {
	t.Helper()

	// run each test consecutively
	for _, test := range tests {
		l := New(test.input)
		for idx, expected := range test.expectedTokens {
			tok := l.NextToken()

			// asserting the token kind is the same
			if tok.Kind != expected.Kind {
				t.Fatalf("tests[%d]: token `Kind` incorrect, expected %v, got %v",
					idx, expected.Kind, tok.Kind)
			}

			// asserting the literal is the same
			if tok.Literal != expected.Literal {
				t.Fatalf("tests[%d]: token `Literal` incorrect, expected %v, got %v",
					idx, expected.Literal, tok.Literal)
			}
		}
	}
}

File: src/test/ast/LexerTest.java
package test.ast;

import org.junit.*;

import org.junit.Assert;
import main.ast.Token;
import main.ast.TokenKind;
import main.ast.Lexer;

import java.util.ArrayList;
import java.util.List;


class LexerTestCase {
    private String input;
    private List<Token> expectedTokens;

    public LexerTestCase(String input, List<Token> expectedTokens) {
        this.input = input;
        this.expectedTokens = expectedTokens;
    }

    public String getInput() {
        return input;
    }

    public List<Token> getExpectedTokens() {
        return expectedTokens;
    }
}

public class LexerTest {
    
    @Test
    public void testNextToken() {
        List<LexerTestCase> tests = new ArrayList<>();
        tests.add(new LexerTestCase(
            "=+(),",
            new ArrayList<Token>() {{
                add(new Token(TokenKind.Bind, "="));
                add(new Token(TokenKind.Plus, "+"));
                add(new Token(TokenKind.LParen, "("));
                add(new Token(TokenKind.RParen, ")"));
                add(new Token(TokenKind.Comma, ","));
            }}
        ));
        tests.add(new LexerTestCase(
            "//STEP010  EXEC PGM=SORT\n" +
            "//SYSOUT   DD SYSOUT=*\n" +
            "//SORTIN   DD DSN=userid.IBMMF.INPUT,DISP=SHR\n" +
            "//SORTOUT  DD DSN=userid.IBMMF.OUTPUT,DISP=SHR\n" +
            "//SYSIN    DD *\n" +
            "  SORT FIELDS=(1,15,ZD,A)\n" +
            "  SUM FIELDS=NONE\n" +
            "/*",
            new ArrayList<Token>() {{
                add(new Token(TokenKind.Executable, "//"));
                add(new Token(TokenKind.Tag, "STEP010"));
                add(new Token(TokenKind.Exec, "EXEC"));
                add(new Token(TokenKind.Parameter, "PGM=SORT"));
                add(new Token(TokenKind.Executable, "//"));
                add(new Token(TokenKind.Tag, "SYSOUT"));
                add(new Token(TokenKind.Dd, "DD"));
                add(new Token(TokenKind.Parameter, "SYSOUT=*"));
                add(new Token(TokenKind.Executable, "//"));
                add(new Token(TokenKind.Tag, "SORTIN"));
                add(new Token(TokenKind.Dd, "DD"));
                add(new Token(TokenKind.Parameter, "DSN=userid.IBMMF.INPUT"));
                add(new Token(TokenKind.Parameter, "DISP=SHR"));
                add(new Token(TokenKind.Executable, "//"));
                add(new Token(TokenKind.Tag, "SORTOUT"));
                add(new Token(TokenKind.Dd, "DD"));
                add(new Token(TokenKind.Parameter, "DSN=userid.IBMMF.OUTPUT"));
                add(new Token(TokenKind.Parameter, "DISP=SHR"));
                add(new Token(TokenKind.Executable, "//"));
                add(new Token(TokenKind.Tag, "SYSIN"));
                add(new Token(TokenKind.Dd, "DD"));
                add(new Token(TokenKind.Aster, "*"));
                add(new Token(TokenKind.SysinMarker, "/*"));
            }}
        ));

        runTest(tests);
    }

    private static void runTest(List<LexerTestCase> tests) {
        // run each test consecutively
        for (LexerTestCase test : tests) {
            Lexer l = new Lexer(test.getInput());
            for (int idx = 0; idx < test.getExpectedTokens().size(); idx++) {
                Token tok = l.nextToken();

                // asserting the token kind is the same
                Assert.assertEquals(String.format("tests[%d]: token `Kind` incorrect, expected %s, got %s",
                                idx, test.getExpectedTokens().get(idx).kind, tok.kind),
                                
                                test.getExpectedTokens().get(idx).kind, 
                                tok.kind
                        );

                // asserting the literal is the same
                Assert.assertEquals(String.format("tests[%d]: token `Literal` incorrect, expected %s, got %s",
                                idx, test.getExpectedTokens().get(idx).literal, tok.literal),
                                
                                test.getExpectedTokens().get(idx).literal,
                                 tok.literal
                        );
            }
        }
    }
}

